flink/src/main/java/org/apache/iceberg/flink/RowDataTaskWriterFactory.java

```diff
@@ -38,37 +40,38 @@
 import org.apache.iceberg.io.OutputFileFactory;
 import org.apache.iceberg.io.TaskWriter;
 import org.apache.iceberg.io.UnpartitionedWriter;
-import org.apache.iceberg.parquet.Parquet;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 
-class RowTaskWriterFactory implements TaskWriterFactory<Row> {
+class RowDataTaskWriterFactory implements TaskWriterFactory<RowData> {
   private final Schema schema;
   private final PartitionSpec spec;
   private final LocationProvider locations;
   private final FileIO io;
   private final EncryptionManager encryptionManager;
   private final long targetFileSizeBytes;
   private final FileFormat format;
-  private final FileAppenderFactory<Row> appenderFactory;
+  private final RowType flinkSchema;
+  private final FileAppenderFactory<RowData> appenderFactory;
 
   private OutputFileFactory outputFileFactory;
 
-  RowTaskWriterFactory(Schema schema,
-                       PartitionSpec spec,
-                       LocationProvider locations,
-                       FileIO io,
-                       EncryptionManager encryptionManager,
-                       long targetFileSizeBytes,
-                       FileFormat format,
-                       Map<String, String> tableProperties) {
+  RowDataTaskWriterFactory(Schema schema,
+                           PartitionSpec spec,
+                           LocationProvider locations,
+                           FileIO io,
+                           EncryptionManager encryptionManager,
+                           long targetFileSizeBytes,
+                           FileFormat format,
+                           Map<String, String> tableProperties) {
     this.schema = schema;
     this.spec = spec;
     this.locations = locations;
     this.io = io;
     this.encryptionManager = encryptionManager;
     this.targetFileSizeBytes = targetFileSizeBytes;
     this.format = format;
-    this.appenderFactory = new FlinkFileAppenderFactory(schema, tableProperties);
+    this.flinkSchema = FlinkSchemaUtil.convert(schema);
```

Comment:
In Spark, we had a bug where Spark may produce a row with a short, which is stored as an int in Iceberg. In a CTAS query, data would actually get passed to Iceberg with the short and we would end up with a ClassCastException. That's why we now pass the dataset schema when creating writers. You might want to watch out for a similar bug.

Reply:
Thanks for the great remanding, I think the bug you mentioned is this one https://github.com/apache/iceberg/pull/999. The flink don't support CTAS but support `INSERT INTO iceberg_table SELECT * from table_2`,  if the table_2 has a TINYINT or SMALLINT, them its `BinaryRowData` queried from `SELECT` will be byte or short,  we also need the raw flink's schema to read the values from `BinaryRowData` (rather than the flink schema converted from iceberg schema), and write those byte or short into integer.  Let me consider how to fix this.

Reply:
This only appears in Spark with a CTAS query because that's the only time that Spark doesn't get a schema back from the table. When Spark has a table schema, it will automatically insert casts to the appropriate types so this problem doesn't happen. I'm not sure if Flink does that, but if it does then you wouldn't need to worry about that bug.